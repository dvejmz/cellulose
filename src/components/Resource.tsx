import React from 'react';

interface ResourceProps {
  name: string;
  classNameId: string;
  value: number;
  max?: number;
  min?: number;
}


const Resource: React.FC<ResourceProps> = (props: ResourceProps) => {
  const getValue = (): number => {
    if (props.min !== undefined && props.value < props.min) {
      return Math.max(props.value, props.min);
    }

    if (props.max !== undefined && props.value > props.max) {
      return Math.min(props.value, props.max);
    }

    return props.value;
  };

  return (
    <div className={`resources__${props.classNameId}`}>
      <span className={`resources__${props.classNameId}-label capitalise`}>{props.name}</span>: <span className={`resources__${props.classNameId}-value`}>{getValue()}</span>
    </div>
  );
};

export default Resource;
