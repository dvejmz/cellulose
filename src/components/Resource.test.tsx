import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Resource from './Resource';

describe('<Resource />', () => {
  let resource: ShallowWrapper;

  beforeEach(() => {
    resource = shallow(<Resource name="Paper" classNameId="paper" value={9001} />);
  });

  it('renders ok', () => {
    expect(resource.exists()).toBeTruthy();
  });

  it('shows resource name', () => {
    expect(resource.find('.resources__paper-label').text()).toBe('Paper')
  });

  it('shows resource value', () => {
    expect(resource.find('.resources__paper-value').text()).toBe('9001');
  });

  it('does not go below specified minimum', () => {
    act(() => {
      resource.setProps({ value: -1000, min: 0 });
    });
    expect(resource.find('.resources__paper-value').text()).toBe('0');
  });

  it('does not go above specified maximum', () => {
    act(() => {
      resource.setProps({ value: 1000, max: 100 });
    });
    expect(resource.find('.resources__paper-value').text()).toBe('100');
  });

  it('shows correct classNames', () => {
    expect(resource.prop('className')).toContain('resources__paper');
    expect(resource.find('.resources__paper-label').exists()).toBeTruthy();
    expect(resource.find('.resources__paper-value').exists()).toBeTruthy();
  });
});
