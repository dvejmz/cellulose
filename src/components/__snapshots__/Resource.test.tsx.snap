// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Resource /> renders ok 1`] = `
<div
  className="resource"
>
  <Counter
    id="paper"
    name="Paper"
    price={0}
    purchaseRate={1}
    quantity={9001}
    quantityUnit="sheets"
  />
</div>
`;
